design-proposals/klip-49-add-source-stream-table-semantic.md

```diff
@@ -0,0 +1,129 @@
+# KLIP 47 - Add Source Stream/Table Semantic 
+
+**Author**: Boyang Chen (@boyang) | 
+**Release Target**: TBD | 
+**Status**: _In Discussion_ | 
+**Discussion**: https://github.com/confluentinc/ksql/pull/7474
+
+**tl;dr:** _Add support for source stream/table. This will give users a read-only view of data that 
+are populated by the upstream Kafka topic, making the data ownership more clear._
+           
+## Motivation and background
+
+Users of KsqlDB have no clear ownership definition around tables or streams they created. In a “normal” database, 
+when a user creates a table the DB creates the corresponding files on disk and owns the table. 
+User may write data into those tables via `INSERT INTO` statements and/or modify/delete existing data 
+via `UPDATE` and `DELETE` statements. However, the table will be not changed by any other means. 
+
+For ksqlDB, the setup is quite differently. Specifically, user could create the table based on 
+the topics stored in Kafka as the source of truth. However, since there is no clear ownership, the defined table 
+could be mutated by using `INSERT` command to populate data into it. The insertion will pollute the original data source, 
+which was supposed to be owned by upstream such as a CDC pipeline that populates the topic.
+Secondly, when calling `CREATE TABLE`, the actual data is neither materialized nor ready for interactive queries to access. 
+To address these gaps, we would like to add:
+
+* Syntax to add read-only stream/table as `CREATE SOURCE STREAM/TABLE` to make the data source as read-only. Any 
+insertion command will be rejected, with an example error response as:
+
+```
+Error: insertion into source stream/table is not allowed
+```   
+
+* In addition, the `CREATE SOURCE TABLE` will become a persistent query instead of just a metadata operation like 
```

Comment:
I think the reasoning for only materializing `SOURCE TABLE` is that we don't want to have all `TABLE` type materialized immediately, but instead taking a first opt-in approach and see user's feedback.

Reply:
It would be a "massive" (breaking?) change to let `CT` start to materialize as people expect it to be a light-weight metadata operation. -- In the long run, my personal proposal is, that we would materialize all tables. But there are also concerns raised by others (ie, cost) that propose to make is configurable via `WITH` clause.

Doing this experiment that is backward compatible should give us more data points to see how we want to move forward.

Reply:
Can we add this discussion to the KLIP if it's not already there?

Reply:
Sounds good